compiler/rustc_mir/src/transform/instcombine.rs

```diff
@@ -41,6 +44,28 @@ impl<'tcx, 'a> InstCombineContext<'tcx, 'a> {
         })
     }
 
+    /// Remove assignments to inhabited ZST places.
+    fn combine_zst(&self, source_info: &SourceInfo, kind: &mut StatementKind<'tcx>) {
+        match kind {
+            StatementKind::Assign(box (place, _)) => {
+                let place_ty = place.ty(self.local_decls, self.tcx).ty;
+                if let Ok(layout) = self.tcx.layout_of(self.param_env.and(place_ty)) {
+                    if layout.is_zst() && !layout.abi.is_uninhabited() {
```

Comment:
Removing assignments of uninhabited ZSTs changed the error output for `ui/consts/const-eval/panic-never-type.rs` and `ui/consts/const-eval/panic-assoc-never-type/panic-assoc-never-type.rs`.

e.g.

```diff
---- [ui] ui/consts/const-eval/panic-assoc-never-type.rs stdout ----
diff of stderr:

15         = note: for more information, see issue #71800 <https://github.com/rust-lang/rust/issues/71800>
16         = note: this warning originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
17
-       error[E0080]: erroneous constant used
+       error: erroneous constant encountered
19        --> $DIR/panic-assoc-never-type.rs:17:13
20         |
21      LL |     let _ = PrintName::VOID;

-          |             ^^^^^^^^^^^^^^^ referenced constant has errors
+          |             ^^^^^^^^^^^^^^^
23
24      error: aborting due to previous error; 1 warning emitted
25

-       For more information about this error, try `rustc --explain E0080`.
27
```

I'm unsure if leaving uninhabited ZSTs alone is the right thing to do or a hack that just happens to fix this case.

Reply:
This change is fine

Reply:
You mean the code is fine, or the error message change would be fine?

Reply:
The change in the error message is totally alright